### PR TITLE
Feat/policy refactor

### PR DESCRIPTION
**Pull request by StewartW**
# Why?

A refactor of how policies (SCP/Tagging) are managed. 

Provides backwards compatibility with the current process whilst enabling an additional folder "adf-policies" that allows for SCPs, Tagging policies (and potentially more) to be defined in a central location. Allowing for policies to be-used across multiple targets. (Currently supports OU-names, but also supports extension going forward) 

 

*Issue #, if available:*

## What?

Description of changes:

Creates a new concept called Policy Campaigns. Which is an object used to orchestrate the updating of policies and their targets. 

Policy Campaigns have a list of Policies that require creating. (A policy is classed as requiring creation when a policy with the same name does not exist). Policies that require updating (A policy is classed as requiring updating when an existing policy has a different content) and policies that require deleting (Any policy that is ADF managed but does not receive an interaction throughout the campaign is considered for deletion) 

Policies themselves do similar logic for targets, and depending on the adf config, a target will either have the default SCP maintained or removed. 

The overall logic is the same for the legacy policy management, the primary difference is that adf-policies is now an optional source for policies. 

---

By submitting this pull request, I confirm that you can use, modify, copy, and
redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/template.yml [hunk 1]
```diff
@@ -18,13 +18,13 @@ Metadata:
     Labels:
       ["adf", "aws-deployment-framework", "multi-account", "cicd", "devops"]
     HomePageUrl: https://github.com/awslabs/aws-deployment-framework
-    SemanticVersion: 3.1.2
+    SemanticVersion: 3.3.9
```
Comment (StewartW) [inline src/template.yml:21]: revert this

**Comment by StewartW** [comment]
Still todo: 

- [ ] Address TODO comments 
- [ ] Update documentation
- [ ]

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+            if organizations_client
+            else boto3.client("organizations")
+        )
+        self.current_targets = self.get_current_targets_for_policy()
+        self.targets_requiring_attachment = {}
+        self.policy_has_changed = policy_has_changed
+        self.targets_not_scheduled_for_deletion = []
+
+    def get_current_targets_for_policy(self):
+        if self.id:
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:129]: Let's flip this one, so we get rid of the indentation and the else statement:

```suggestion
        if not self.id:
          return []
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+                current_targets = [
+                    t.get("TargetId")
+                    for t in self.organizations_client.list_targets_for_policy(
+                        PolicyId=self.id
+                    ).get("Targets")
+                ]
+                return current_targets
+            except self.organizations_client.exceptions.AccessDeniedException as e:
+                LOGGER.critical(
+                    f"Error fetching targets for policy {self.name} {self.id}: Access Denied"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:140]:
```suggestion
                    "Error fetching targets for policy %s %s: Access Denied",
                    self.name,
                    self.id,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+                )
+                LOGGER.error(e)
+                raise OrganizationPolicyException(
+                    f"Error fetching targets for policy {self.name} {self.id}: Access Denied"
+                ) from e
+            except self.organizations_client.exceptions.AWSOrganizationsNotInUseException as e:
+                LOGGER.critical(
+                    f"Error fetching targets for policy {self.name} {self.id}: Organizations not in use"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:148]:
```suggestion
                    "Error fetching targets for policy %s %s: Organizations not in use",
                    self.name,
                    self.id,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+                )
+                LOGGER.error(e)
+                raise OrganizationPolicyException("Organizations not in use") from e
+            except self.organizations_client.exceptions.InvalidInputException as e:
+                LOGGER.critical(
+                    f"Error fetching targets for policy {self.name} {self.id}: Invalid Input"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:154]:
```suggestion
                    "Error fetching targets for policy %s %s: Invalid Input",
                    self.name,
                    self.id,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+                )
+                LOGGER.error(e)
+                raise OrganizationPolicyException(
+                    f"Error fetching targets for policy {self.name} {self.id}: Invalid Input"
+                ) from e
+            except self.organizations_client.exceptions.ServiceException as e:
+                LOGGER.critical(
+                    f"Error fetching targets for policy {self.name} {self.id}: Service Exception "
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:162]:
```suggestion
                    "Error fetching targets for policy %s %s: Service Exception ",
                    self.name,
                    self.id,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+                )
+                LOGGER.error(e)
+                raise OrganizationPolicyException(
+                    f"Error fetching targets for policy {self.name} {self.id}: Service Exception "
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:166]:
```suggestion
                    "Error fetching targets for policy %s %s: Service Exception ",
                    self.name,
                    self.id,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+                ) from e
+            except self.organizations_client.exceptions.TooManyRequestsException as e:
+                LOGGER.critical(
+                    f"Error fetching targets for policy {self.name} {self.id}: Too Many Requests "
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:170]:
```suggestion
                    "Error fetching targets for policy %s %s: Too Many Requests ",
                    self.name,
                    self.id,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+                )
+                LOGGER.error(e)
+                raise OrganizationPolicyException(
+                    "Too Many Requests to Organizations API"
+                ) from e
+            except Exception as e:
+                LOGGER.critical(
+                    f"Error fetching targets for policy {self.name} {self.id}: Unexpected exception"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:178]:
```suggestion
                    "Error fetching targets for policy %s %s: Unexpected exception",
                    self.name,
                    self.id,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+
+        for target_id in targets_to_detach:
+            LOGGER.info("Detaching policy (%s) from target (%s)", self.name, target_id)
+            try:
+                self.organizations_client.detach_policy(
+                    PolicyId=self.id, TargetId=target_id
+                )
+            except self.organizations_client.exceptions.AccessDeniedException as e:
+                LOGGER.critical(
+                    f"Error detaching policy {self.name} ({self.id}) from target {target_id}: Access Denied"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/organization_policy_campaign.py:231]: Can you fix the other logs to use the `%s` style too?

File: src/template.yml [hunk 11]
```diff
@@ -18,13 +18,13 @@ Metadata:
     Labels:
       ["adf", "aws-deployment-framework", "multi-account", "cicd", "devops"]
     HomePageUrl: https://github.com/awslabs/aws-deployment-framework
-    SemanticVersion: 3.2.0
+    SemanticVersion: 3.2.1
```
Comment (sbkok) [inline src/template.yml:21]:
```suggestion
    SemanticVersion: 3.2.0
```

File: src/template.yml [hunk 12]
```diff
@@ -18,13 +18,13 @@ Metadata:
...
     SourceCodeUrl: https://github.com/awslabs/aws-deployment-framework
 
 Mappings:
   Metadata:
     ADF:
-      Version: 3.2.0
+      Version: 3.2.1
```
Comment (sbkok) [inline src/template.yml:27]:
```suggestion
      Version: 3.2.0
```

File: docs/admin-guide.md [hunk 13]
```diff
@@ -730,6 +730,67 @@ Once you have enabled all features within your Organization, ADF can manage and
...
+
+A new feature of ADF, that gives you the ability to define a policy in a
+single location, and apply it to multiple targets.
+
+### Enabling the new version
+
+Because of the difference in this approach to applying policies, it is not
+currently the default method and will have to be enabled. In order to enable it,
+you have to update your serverlessrepo stack in the organizational root account
+and set the parameter `EnablePolicyV2' to "TRUE". Once the stack has redeployed,
```
[the remark this answers is not in the document]
Comment (StewartW) [reply docs/admin-guide.md:756]: Happy to postpone to V4.0.0.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/main.py [hunk 14]
```diff
@@ -204,54 +199,37 @@ def _store_extension_parameters(parameter_store, config):
...
-        account_id=account_id
-    )
+    organizations = Organizations(role=boto3, account_id=account_id)
     ou_id = organizations.get_parent_info().get("ou_parent_id")
 
     account_path = organizations.build_account_path(
-        ou_id,
-        [],  # Initial empty array to hold OU Path,
-        cache
+        ou_id, [], cache  # Initial empty array to hold OU Path,
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/main.py:213]: Can you revert this change, as the comment is linked to the empty list. Not the cache or all three arguments together?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+            glob.iglob(
+                f"{self.adf_policies_dir}/{policy}/*.json",
+                recursive=True,
+            )
+        )
+        return [f.replace(f"{self.adf_policies_dir}", ".") for f in _files]
+
+    @staticmethod
+    def _trim_scp_file_name(policy):
+        return policy[1:][:-8] if policy[1:][:-8] == "/" else policy[2:][:-9]
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py:57]: Can you add comments here, explaining why 1 and -8 were chosen?
Plus what is the other use case, with 2 and -9?

The function is called trim, while it seems to only return a "/" in the first use case.
Is this correct?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+    @staticmethod
+    def _trim_tagging_policy_file_name(policy):
+        return policy[1:][:-19] if policy[1:][:-19] == "/" else policy[2:][:-20]
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py:61]: Same here, can you add comments here too. Preferably with examples?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        ) as policy:
+            return json.dumps(json.load(policy))
+
+    def apply(
+        self, organizations, parameter_store, config
+    ):  # pylint: disable=R0912, R0915
+        status = organizations.get_organization_info()
+        if status.get("feature_set") != "ALL":
+            LOGGER.info(
+                "All Features are currently NOT enabled for this Organization, "
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py:85]: If we are tasked to apply an SCP or Tagging Policy, maybe it would be better to raise this as a warning or even a fatal error. As it might go unnoticed otherwise. What do you think?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+                )
+                pass
+            for _policy in _legacy_policies:
+                LOGGER.info("Loading policy: %s", _policy)
+                proposed_policy = json.loads(Organizations.get_policy_body(_policy))
+
+                path = (
+                    OrganizationPolicy._trim_scp_file_name(_policy)
+                    if policy == "scp"
+                    else OrganizationPolicy._trim_tagging_policy_file_name(_policy)
```
[the remark this answers is not in the document]
Comment (sbkok) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/organization_policy_v2.py:173]: Never mind, this is in the same file. I didn't realize that before.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/tests/test_organizational_policy_campaigns.py [hunk 19]
```diff
@@ -0,0 +1,1159 @@
...
+            policy.set_targets(
+                [policy_campaign.get_target(t) for t in _policy.get("Targets")]
+            )
+            self.assertEqual(1, len(policy.targets_requiring_attachment))
+
+        policy_campaign.apply()
+        stubber.assert_no_pending_responses()
+
+
+class SadTestCases(unittest.TestCase):
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/tests/test_organizational_policy_campaigns.py:907]: Can you add a test case that will test that we throw an error when AWS Organizations is not configured with all features that we require for Tagging and Service Control Policies?

**Review by sbkok — CHANGES_REQUESTED** [review]
A few comments still there from the previous review + a few new ones.
Could you look at these via the Files changed tab? That should only show the comments that still apply.

File: docs/admin-guide.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (StewartW) [reply docs/admin-guide.md:891]: Happy to get your thoughts on this, because the new V2 approach is 100% backwards compatible with old definitions

File: Makefile [hunk 20]
```diff
@@ -214,12 +214,12 @@ verify_tooling: .venv
 			exit 1 \
 		); \
 	)
-	@( \
-		sed --version &> /dev/null || ( \
-			echo '$(CLR_RED)Sed is not installed!$(CLR_END)' && \
-			exit 1 \
-		); \
-	)
+	# @( \
```
Comment (StewartW) [inline Makefile:217]: This isn't reliable on mac. Is there a better way to do this @sbkok ?

**Comment by StewartW** [comment]
Closing to favour the v4.0.0 compatible version
